### PR TITLE
[RISCV] Add vector and vector crypto to SiFiveP400 scheduler model

### PR DESCRIPTION
**Pull request by michaelmaitland**
The SiFiveP400 scheduler model did not support vector or vector crypto. With
    the addition of the sifive-p470 processor, this model needs to support these
    extensions.

    The processors who use this model but do not have vector or vector crypto
    will never produce these instructions, so there is no impact to these
    processors.

    This change is stacked on #102022

    Co-authored-by: Min Hsu <min.hsu@sifive.com>

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td [hunk 1]
```diff
@@ -45,6 +154,13 @@ defvar SiFiveP400FloatArith  = SiFiveP400FEXQ0;
 defvar SiFiveP400F2I      = SiFiveP400FEXQ0;
 def SiFiveP400FloatDiv    : ProcResource<1>;
 
+// Vector pipeline
+def SiFiveP400VEXQ0        : ProcResource<1>;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td:158]: `EXQ0` means `execution queue 0`? But it seems there is no second unit and all arithmetic operations are executed on this unit (not a quene?).

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td:158]: There is an issue queue in front of the ALU. https://www.cnx-software.com/wp-content/uploads/2022/11/SiFive-P470-detailed-pipeline-overview.png

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td:158]: So there is 1 vector pipeline? Is it OK to model VLD/VST/VDiv/VFloatDiv as standalone resources?

**Comment by camel-cdr** [comment]
Are the vrgather.vv numbers correct? Usually LMUL>1 vrgather is implemented by aplying a LMUL=1 vrgather LMUL^2 times. Since the LMUL=1 vrgather.vv takes a single cycle, I would've expected a 4 cycle LMUL=2 vrgather.vv instead of 12.
12 cycles is less than one element per cycle (128/32*2=8).

Oh, and did you use upstream llvm-exegesis? I didn't managed to get it to work last time I tried.

File: llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td (hunk 1 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td:158]: There is an issue queue in front of the vector ALU (purple). Vector loads go through the load/store dispatch buffer (green). Vector divisions are non blocking to other vector ALU operations.

We think it is okay to model VLD/VST/VDiv/VFloatDiv as standalone resources.

**Comment by michaelmaitland** [comment]
@camel-cdr

> Are the vrgather.vv numbers correct?

According to llvm-exegisis, our micro-architecture spec, and llvm-mca reports, we believe these numbers are correct.

> Usually LMUL>1 vrgather is implemented by applying a LMUL=1 vrgather LMUL^2 times. Since the LMUL=1 vrgather.vv takes a single cycle, I would've expected a 4 cycle LMUL=2 vrgather.vv instead of 12.

We don't think using the approach you present here agrees with what we see in llvm-exegisis. We think the approach we take is a good representation.

> Oh, and did you use upstream llvm-exegesis? I didn't managed to get it to work last time I tried.

This is a question for @mshockwave.

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Pull request merged**